### PR TITLE
Fill type info assist

### PR DESCRIPTION
**Pull request by omerbenamram**
I was looking into #1965,

I got the simplest of tests passing:
```
"fn f() { let a<|>: _ = 1; }" == "fn f() { let a<|>: i32 = 1; }",
```

This basically scans the text range after the colon for the placeholder type, and calls `analyzer.type_of` like the previous implementation, plus it doesn't bail immediatly after seeing a colon.

I'd like some help with is understanding why I'm getting the `Unknown` type when trying to retrieve type information for `vec![1,2,3]`.

It clearly works when on the IDE, so RA can infer this correctly:
image

What am I missing?

This is probably a pretty messy implementation - we could move `followed_by_placeholder_ty` outside and so on, and perhaps the traversal is not optimal - so comment/nits on the implementation details are more than welcome!

### CODE REVIEW

**Comment by edwin0cheng** [comment]
> 'd like some help with is understanding why I'm getting the Unknown type when trying to retrieve type information for vec![1,2,3].

It is because when lowering, we do not store the `expr_id` of the actual macro_cal ast node, we only store all expanded exprs inside `BodySourceMap` such that the following line will return `None`:

https://github.com/rust-analyzer/rust-analyzer/blob/d6223253b628b279f9ddae8f83f7173d01f6b32c/crates/ra_hir/src/source_binder.rs#L216

Maybe we could try to expand the `Ast::MacroCall` before/inside calling `SourceAnalyzer::type_of` ?

**Comment by omerbenamram** [comment]
@edwin0cheng I'm not sure thats a macro problem - I've also tried `"fn f() { let a<|>: Vec<_> = Vec::<i32>::new()"` - and it still returns `Unknown`. Also - the call to `analyzer.type_of` does return a `ty` - it simply is unknown.

puttind a `dbg!(&expr, &ty);` after line 91 at `add_explicit_type.rs` yields:
```
[crates\ra_assists\src\assists\add_explicit_type.rs:93] &expr = CallExpr(
    CallExpr {
        syntax: CALL_EXPR@[25; 42)
          PATH_EXPR@[25; 40)
            PATH@[25; 40)
              PATH@[25; 35)
                PATH_SEGMENT@[25; 35)
                  NAME_REF@[25; 28)
                    IDENT@[25; 28) "Vec"
                  TYPE_ARG_LIST@[28; 35)
                    COLONCOLON@[28; 30) "::"
                    L_ANGLE@[30; 31) "<"
                    TYPE_ARG@[31; 34)
                      PATH_TYPE@[31; 34)
                        PATH@[31; 34)
                          PATH_SEGMENT@[31; 34)
                            NAME_REF@[31; 34)
                              IDENT@[31; 34) "i32"
                    R_ANGLE@[34; 35) ">"
              COLONCOLON@[35; 37) "::"
              PATH_SEGMENT@[37; 40)
                NAME_REF@[37; 40)
                  IDENT@[37; 40) "new"
          ARG_LIST@[40; 42)
            L_PAREN@[40; 41) "("
            R_PAREN@[41; 42) ")"
        ,
    },
)
[crates\ra_assists\src\assists\add_explicit_type.rs:93] &ty = Type {
    krate: CrateId(
        0,
    ),
    ty: InEnvironment {
        environment: TraitEnvironment {
            predicates: [],
        },
        value: Unknown,
    },
}
```

**Comment by edwin0cheng** [comment]
It is because our test fixture do not import anything (include std),  so it will not works in such case.
However, I try to build and install your PR as actual VS Code Extension, it works: 

it_works

**Comment by omerbenamram** [comment]
@edwin0cheng I was silly not to actually try this! I've added an example that doesn't need std and it passes. Thanks for the help :)

**Comment by omerbenamram** [comment]
@edwin0cheng I've checked it now - and it seems to fail with the macro `vec!`, but work with `Vec`.
So this probably related to the issue that you've previously mentioned.

**Comment by omerbenamram** [comment]
@edwin0cheng I've tried doing something like this before calling `analyzer.type_of`:

```
    if let Some(macro_call) = ast::MacroCall::cast(expr.syntax().clone()) {
        let token = hir::InFile::new(ctx.frange.file_id.into(), macro_call);
        let expansion = analyzer.expand(db, token.as_ref())?;
        dbg!(expansion);
    };
```
It does seem to hit and print:

```
[crates\ra_assists\src\assists\add_explicit_type.rs:101] expansion = Expansion {
    macro_call_id: MacroCallId(
        0,
    ),
}
```

But it won't be inferred if I simply check expr after it.
Do I need to somehow ask it to infer the macro?

Also - I'm using a lot of the code found in the `hover` action as reference, it seems to  handle all of those cases, and - evidently it gets it right a lot of the times, I wonder if theres any way we can share some of the code with `ra_ide`?

**Comment by edwin0cheng** [comment]
> @edwin0cheng I've tried doing something like this before calling analyzer.type_of:

IMHO, maybe just leave it as is and solve it in another PR ? Fixing it seem like out of scope of this PR and it involve some design decision in `type_of` and `lowering` which I am not familiar to.

**Comment by omerbenamram** [comment]
@edwin0cheng I'm fine with fixing it later - it's somewhat beyond my current comprehension of the code base aswell. 
I'll wait for @matklad to look over this and see how to proceed from here :)

File: crates/ra_assists/src/assists/add_explicit_type.rs [hunk 1]
```diff
@@ -25,31 +30,85 @@ pub(crate) fn add_explicit_type(ctx: AssistCtx<impl HirDatabase>) -> Option<Assi
     let stmt = ctx.find_node_at_offset::<LetStmt>()?;
```
Comment (matklad) [inline crates/ra_assists/src/assists/add_explicit_type.rs]: I think let and `_` cases are sufficiently different that there's little sense to handle both in the same function. There are cases like `xs.iter().collect::<_>()`, where there are no `let` at all. 

I think we should add a second `pub(crate)` function in this file

```rust
pub(crate) fn fill_placeholder_type(ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
   let placeholder = cts.find_node_at_offset::<ast::Placeholder>()?;
   ...

```

**Comment by matklad** [comment]
Yeah, it's a good point that `type_of` currently doesn't work with macros, but this needs to be fixed on the deeper level. For the time being, let's ignore macros altogether.

File: crates/ra_assists/src/assists/add_explicit_type.rs [hunk 2]
```diff
@@ -25,31 +30,85 @@ pub(crate) fn add_explicit_type(ctx: AssistCtx<impl HirDatabase>) -> Option<Assi
...
         return None;
     }
 
     ctx.add_assist(AssistId("add_explicit_type"), "add explicit type", |edit| {
         edit.target(pat_range);
-        edit.insert(name_range.end(), format!(": {}", ty.display(db)));
+
+        match assist_type {
+            AddExplicitTypeAssist::PlaceHolder(range) => {
+                edit.replace(range, format!(": {} ", ty.display(db)))
```
Comment (matklad) [inline crates/ra_assists/src/assists/add_explicit_type.rs]: Ah, so this doesn't really *fill* the placeholder, but replaces the whole type! Heh, that's neat, I haven't realised that this is poissible. 

But I think we should try to replace placeholder specifically. The original issue mentioned "walking type and pattern in lockstep", but I wonder if a more direct solution is possible. 

So, we start with `ast::BindPat`. That can be upcast via `.into` to `ast::Pat`, and that can be used with `SourceAnalyzer::type_of_pat`, which should give a type of this `_`

**Comment by matklad** [comment]
Closing due to inactivity!

**Pull request closed without merging**